include/rx/ranges.hpp

```diff
@@ -1187,142 +1244,66 @@ template <class... Inputs>
-            size_t i = 0;
-            while (true) {
-                if (RX_LIKELY(!inner.at_end())) {
-                    storage[i++] = inner.get();
-                    if (RX_LIKELY(i != N)) {
-                        inner.next();
-                    } else {
-                        return;
-                    }
-                } else {
-                    return;
-                }
-            }
-        }
-    };
-
-    template <class R>
-    [[nodiscard]] constexpr auto operator()(R&& input) const {
-        using Inner = get_range_type_t<R>;
-        return Range<Inner>{as_input_range(std::forward<R>(input))};
-    }
-};
-
 /*!
-    @brief Produce sequential groups of exactly N elements, where N is only known at runtime.
+    @brief Produce sequential groups of exactly N elements.
 
-    The output type is `std::vector<N>` const reference, where T is the output type of the inner
-    range (without cvref-qualifiers), and the size of the result is always N.
+    The output type is a range of exactly N element.
 
-    If the input produces a number of inputs that is not divisible by N, those elements will be
+    If the input produces a number of inputs that is not divisible by N, elements at the end will be
     skipped.
 */
-template <>
-struct in_groups_of_exactly<> {
-    size_t n;
+struct in_groups_of_exactly {
+    const size_t n;
 
     constexpr explicit in_groups_of_exactly(size_t n) noexcept : n(n) {}
 
     template <class R>
     struct Range {
         using element_type = remove_cvref_t<typename R::output_type>;
-        using output_type = const std::vector<element_type>&;
+        using output_type = remove_cvref_t<decltype(std::declval<R>() | take(1))>;
         static constexpr bool is_finite = is_finite_v<R>;
-        static constexpr bool is_idempotent = true; // we have internal storage
+        static constexpr bool is_idempotent = true; // ... but we output potentially non-idempotent
+                                                    // ranges.
 
         R inner;
         const size_t n;
-        std::vector<element_type> storage;
+        RX_OPTIONAL<output_type> storage;
 
         constexpr explicit Range(R inner, size_t n) : inner(std::move(inner)), n(n) {
-            storage.reserve(n);
-            fill_group();
+            next();
         }
 
 
-        [[nodiscard]] constexpr output_type get() const noexcept {
-            return storage;
+        [[nodiscard]] constexpr output_type get() const& noexcept {
+            return *storage;
+        }
+
+        [[nodiscard]] constexpr output_type get() && noexcept {
+            return std::move(*storage);
         }
 
         [[nodiscard]] constexpr bool at_end() const noexcept {
-            return inner.at_end();
+            return bool(!storage);
         }
 
         constexpr void next() noexcept {
-            inner.next();
-            fill_group();
+            storage.reset();
+            if (!inner.at_end()) {
```

Comment:
`LIKELY(!inner.at_end())` ?

Reply:
Well, since there is no `else` block, it's very unlikely (hehe) to make a difference. Still, might as well.